##**Introdução**

Estamos chegando na reta final, e este módulo dará início ao desenvolvimento de um projeto prático que passará por todo o workflow em Data Science, do entendimento do problema, tratamento e análise dos dados até a proposta de solução utilizando Machine Learning.

Vamos trabalhar com dados da COVID-19 do hospital Sírio Libanês, focando este módulo no tratamento e análise de dados para entender profundamente o problema que estamos lidando e propor possíveis soluções.

A partir de agora, tudo que foi aprendido ao longo destes meses será colocado em prática para ajudar a propor soluções para problemas reais de um hospital. 

##**Dados**
Os dados foram disponiblizados no [Kaggle](https://www.kaggle.com/S%C3%ADrio-Libanes/covid19) pela equipe de *Data Intelligence* do hospital Sírio-Libanês. 


Os dados estão organizados em janelas: 0-2 horas, 2-4 horas, 4-6 horas, 6-12 horas e > 12 horas, indicando em quantas horas ocorreu a admissão na UTI após a entrada dos pacientes no hospital. Além disso, o dataset conta com as seguintes informações: Informações demográficas do paciente, Doenças anteriores agrupadas de pacientes, Resultados de exames de sangue e Sinais vitais. 

##**Objetivo**


**Problema:** A pandemia de COVID-19 causou uma sobrecarga nos sistemas de saúde, que afetou a disponibilidade de leitos de UTI nos hospitais. Portanto, é de extrema importância a obtenção de dados precisos para prever e preparar os sistemas de saúde e evitar colapsos, definidos pela necessidade de leitos de UTI acima da capacidade. Essas previsões podem ser realizadas utilizando os  dados clínicos individuais.
Então, esse problema proposto envolve duas tarefas:

**Tarefa 01**
Prever admissão na UTI de casos confirmados de COVID-19.
Com base nos dados disponíveis, é viável prever quais pacientes precisarão de suporte em unidade de terapia intensiva?
O objetivo é fornecer aos hospitais terciários e trimestrais a resposta mais precisa, para que os recursos da UTI possam ser arranjados ou a transferência do paciente possa ser agendada.

**Tarefa 02**
Prever NÃO admissão à UTI de casos COVID-19 confirmados.
Com base nos dados disponíveis, é possível prever quais pacientes precisarão de suporte de unidade de terapia intensiva?
O objetivo é fornecer aos hospitais locais e temporários uma resposta boa o suficiente, para que os médicos de linha de frente possam dar alta com segurança e acompanhar remotamente esses pacientes.



##**Importando as bibliotecas**

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.linear_model import LogisticRegression 
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')

In [2]:
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

In [ ]:
!pip install xlrd

In [ ]:
!pip install openpyxl

##**Importando os dados**

In [3]:
dados = pd.read_excel('https://github.com/vqrca/bootcamp_alura_aulas/blob/main/Modulo_4/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true')
dados.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

##**Aula 2**

In [ ]:
dados.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

Analisando a última coluna: `dados["ICU"]`, que indica a entrada na UTI:

In [ ]:
dados["ICU"]

0       0
1       0
2       0
3       0
4       1
       ..
1920    0
1921    0
1922    0
1923    0
1924    0
Name: ICU, Length: 1925, dtype: int64

Vamos ver quantas pessoas entraram na UTI:

In [ ]:
dados["ICU"].value_counts()

0    1410
1     515
Name: ICU, dtype: int64

ICU: 1410 pacientes entraram na UTI e 515 não? 

Todas as outras colunas são informações sobre os pacientes.

##**Aula 3**

In [ ]:
x_columns = dados.describe().columns

dados_limpos = dados.dropna()

In [ ]:
y = dados_limpos["ICU"]
x = dados_limpos[x_columns].drop(["ICU"], axis=1)

In [ ]:
modelo = LogisticRegression()
modelo.fit(x, y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
modelo.predict([x.iloc[4]])

array([1])

In [ ]:
x.iloc[4]

PATIENT_VISIT_IDENTIFIER           2.000000
AGE_ABOVE65                        0.000000
GENDER                             0.000000
DISEASE GROUPING 1                 0.000000
DISEASE GROUPING 2                 0.000000
                                     ...   
BLOODPRESSURE_SISTOLIC_DIFF_REL   -0.343258
HEART_RATE_DIFF_REL               -0.576744
RESPIRATORY_RATE_DIFF_REL         -0.695341
TEMPERATURE_DIFF_REL              -0.505464
OXYGEN_SATURATION_DIFF_REL        -0.900129
Name: 14, Length: 228, dtype: float64

In [ ]:
y.iloc[4]

1

In [ ]:
sum(modelo.predict(x) == y) #para saber o número de acertos

497

In [ ]:
sum(modelo.predict(x) != y)  #para saber o número de erros

48

In [ ]:
accuracy = sum(modelo.predict(x) == y) / len(y) * 100 #quantos % de acerto
accuracy

91.19266055045871

##**Aula 4**

In [ ]:
y.value_counts()

0    281
1    264
Name: ICU, dtype: int64

**Melhorando as análises de ML:**

Pois estávamos testando o modelo com x e y e ele simplesmente repete.

Chutar a maioria dá mais chance de acerto...

In [ ]:
modelo = DummyClassifier("most_frequent") #utiliza a resposta mais frequente 
modelo.fit(x, y)
sum(modelo.predict(x) == y) / len(y) * 100

51.559633027522935

Acurácia é a soma dos acertos da classificação dividida pelo total de elementos testados:

In [ ]:
accuracy_score(y, modelo.predict(x))

0.5155963302752293

Definindo o Seed para deixar as análises reprodutíveis: 

In [ ]:
import numpy as np

np.random.seed(73246)

Separando dados de treino e teste:

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y)

In [ ]:
print(len(x_train), len(y_train)) #número de dados usados no treino 
print(len(x_test), len(y_test)) #número de dados usados no teste

408 408
137 137


Novo modelo: LogisticRegression

In [ ]:
modelo = LogisticRegression()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.8248175182481752

In [ ]:
y_train.value_counts(normalize=True)

0    0.514706
1    0.485294
Name: ICU, dtype: float64

In [ ]:
y_test.value_counts(normalize=True)

0    0.518248
1    0.481752
Name: ICU, dtype: float64

Testar em cima dos dados de treino cria um viés pra quem está sendo treinado, por isso os dados de treino e teste têm que estar distribuídos de forma adequada. Essa distribuição é garantida através do `stratify `. Já a reproducibilidade é garantida através do `seed`.

##**Aula 5**

**Entendo os dados da forma correta:**

- Cada paciente é identificado pela primeira coluna. São 385 pacientes, que estão repetidos em várias linhas (5 linhas para cada paciente).

- Eles agruparam os dados por janelas de tempo (coluna window)

- Foi feito um agrupamento na hora do exame.

- Portanto, os dados estão desnormalizados!

In [227]:
dados.head(10)

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1
5                         1            1  ...       0-2    1
6                         1            1  ...       2-4    1
7                         1            1  ...       4-6    1
8                         1            1  ...      6-12    1
9                         1            1  ...  ABOVE_12    1

[10 rows x 231 columns]

No `dados.describe` nós temos 1925 pacientes, mas na verdade temos 385 pacientes. Esse conjunto de dados está desnormalizado, pois temos várias linhas representando uma entidade única, que no caso é um paciente.
Então, para analisar cada paciente, temos que analisar várias linhas.
Provavelmente o hospital disponibilizou os dados assim, pois quanto maior o número de dados: menor a privacidade! 

Eles agruparam os dados de cada paciente por janelas de tempo (coluna `window`):
- 0 - 2h
- 2 - 4h
- 4 - 6h
- 6 - 12h
- above 12h

Isso conta a história do paciente nessas janelas de tempo. 

Eles colocaram a média de todos os exames de sangue realizados.

Então, temos dos pontos importantes: 

1 - Se pensarmos bem é necessário excluir todas as linhas onde o ICU equivale a 1, pois pode ser que a pessoa foi primeiro para UTI e depois fez os exames de sangue. Então, a mediação de sangue fica como valor de X, o que não faz sentido aparecer a mediação de sangue depois do paciente ter ido para UTI. A gente só sabe essa informação depois da pessoa ir para UTI e nós queremos fazer a pergunta da previsão ANTES da pessoa ter ido para UTI. 

Toda linha de ICU que tem 1 precisa ser descartada! 

2 - Quais colunas vamos usar??? 

- Usar tudo até Window e então tentamos descobrir se a pessoa vai para UTI na próxima janela. Então, pegamos o ICU da próxima janela. Para isso é necessário filtrar somente as linhas com 0, usar o x de uma linha e o y da linha seguinte. Então, com os dados das duas primeiras horas será que o paciente vai para UTI na próxima hora??? e assim por diante nas próximas horas. 

O que medimos frequentemente não costuma ter um salto de uma medição para outra. Numa série temporal o melhor chute é sempre o valor anterior...

Podemos pegar uma leitura de cada paciente:



Quando fazemos um groupby o grupo vira um index


Começando do zero:

In [420]:
dados = pd.read_excel('https://github.com/vqrca/bootcamp_alura_aulas/blob/main/Modulo_4/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true')
dados.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

Vamos definir as features com valores contínuos, que são as medições colunas 13 a -2:

In [421]:
dados.iloc[:, 13:-2].columns

Index(['ALBUMIN_MEDIAN', 'ALBUMIN_MEAN', 'ALBUMIN_MIN', 'ALBUMIN_MAX',
       'ALBUMIN_DIFF', 'BE_ARTERIAL_MEDIAN', 'BE_ARTERIAL_MEAN',
       'BE_ARTERIAL_MIN', 'BE_ARTERIAL_MAX', 'BE_ARTERIAL_DIFF',
       ...
       'HEART_RATE_DIFF', 'RESPIRATORY_RATE_DIFF', 'TEMPERATURE_DIFF',
       'OXYGEN_SATURATION_DIFF', 'BLOODPRESSURE_DIASTOLIC_DIFF_REL',
       'BLOODPRESSURE_SISTOLIC_DIFF_REL', 'HEART_RATE_DIFF_REL',
       'RESPIRATORY_RATE_DIFF_REL', 'TEMPERATURE_DIFF_REL',
       'OXYGEN_SATURATION_DIFF_REL'],
      dtype='object', length=216)

Agora vamos criar uma função para o preenchimento das linhas das features contínuas sem valores, usando os métodos `bfill` e `ffill`.

Além disso, vamos definir as features categóricas e os dados de saída:

- categóricas:  `dados.iloc[:, :13]` (São as colunas com valores de 0 e 1)
- saída: `dados.iloc[:, -2:]`

Posteriormente, os dados serão concatenados `pd.concat([features_categoricas, features_continuas, saida]`

In [422]:
def preenche_tabela(dados):
    features_continuas_colunas = dados.iloc[:, 13:-2].columns
    features_continuas = dados.groupby("PATIENT_VISIT_IDENTIFIER", as_index=False)[features_continuas_colunas].fillna(method='bfill').fillna(method='ffill')
    features_categoricas = dados.iloc[:, :13]
    saida = dados.iloc[:, -2:] #:, todas as linhas
    dados_finais = pd.concat([features_categoricas, features_continuas, saida], ignore_index=True,axis=1) #definir eixo=1 
    dados_finais.columns = dados.columns
    return dados_finais

In [423]:
dados_limpos = preenche_tabela(dados)
dados_limpos.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

In [424]:
dados_limpos.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

Agora vemos que todos os valores de count estão próximos, exceto nas colunas de doenças. Há algum paciente que não tem as doenças e algumas outras coisas. Então, vamos precisar dar um `dropna`:

In [425]:
x_columns = dados.describe().columns

dados_limpos = dados_limpos.dropna()

Agora vamos retirar pacientes que entraram na UTI nas duas primeiras horas:

In [426]:
a_remover = dados_limpos.query("WINDOW=='0-2' and ICU==1")['PATIENT_VISIT_IDENTIFIER'].values
dados_limpos = dados_limpos.query("PATIENT_VISIT_IDENTIFIER not in @a_remover")
dados_limpos.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1760.000000  ...  1760.000000
mean                 192.818182  ...     0.201705
std                  110.637724  ...     0.401387
min                    0.000000  ...     0.000000
25%                   97.750000  ...     0.000000
50%                  191.500000  ...     0.000000
75%                  289.250000  ...     0.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

Agora precisamos definir qua janela pegar!

Vamos simplificar e pegar a primeira janela: quando a pessoa entra (nas duas primeiras horas). 

Porém, precisamos saber se o paciente foi ou não para UTI:

Para isso foi criada uma função que marca a primeira linha e devolve a primeira linha, sendo todos os pacientes (uma linha por paciente) com a informação se ele foi para UTI (=1) e se ele não foi para UTI (=0).

In [427]:
def prepare_window(rows):
    if(np.any(rows["ICU"])):
        rows.loc[rows["WINDOW"]=="0-2", "ICU"] = 1
    return rows.loc[rows["WINDOW"] == "0-2"]

"""Essa função recebe as linhas (rows), que são as linhas de um paciente e se qualquer
linha ICU do paciente for = 1  vamos pegar a primeira linha desse paciente e marcar que
ele foi para UTI. Então, a linha onde está a janela 0-2 será localizada e retornada através de 
rows.loc[rows["WINDOW"] == "0-2" """

dados_limpos = dados_limpos.groupby("PATIENT_VISIT_IDENTIFIER").apply(prepare_window)
dados_limpos.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
PATIENT_VISIT_IDENTIFIER                                            ...            
0                        0                          0            1  ...    0-2    1
2                        10                         2            0  ...    0-2    1
3                        15                         3            0  ...    0-2    0
4                        20                         4            0  ...    0-2    0
5                        25                         5            0  ...    0-2    0

[5 rows x 231 columns]

Separar os dados em x e y, para aplicar o modelo:

In [428]:
y = dados_limpos["ICU"]
x = dados_limpos[x_columns].drop(["ICU"], axis=1)

Adicionando o seed, para que o `testsplit` ocorra da mesma forma:

In [429]:
np.random.seed(79230)

In [430]:
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y)

DummyClassifier

In [431]:
modelo = DummyClassifier()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.5113636363636364

LogisticRegression

In [432]:
modelo = LogisticRegression()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.7159090909090909

##**Desafios**

###**Desafio 01: Se você não conhece o Kaggle, explore um pouco os datasets, competições e discussões para verificar o quão rico é esta plataforma.**

✅ 

Conta criada: https://www.kaggle.com/valquiriaalencar

###**Desafio 02: Leia a descrição do problema proposta pelo Sírio Libanês, tentando entender qual é o real problema. Antes de continuar para as próximas aulas, debata no Discord suas conclusões sobre o projeto.**

**Real problema:** A pandemia de COVID-19 causou uma sobrecarga nos sistemas de saúde, que afetou a disponibilidade de leitos de UTI nos hospitais. Portanto, é de extrema importância a obtenção de dados precisos para prever e preparar os sistemas de saúde e evitar colapsos, definidos pela necessidade de leitos de UTI acima da capacidade. Essas previsões podem ser realizadas utilizando os  dados clínicos individuais.
Então, esse problema proposto envolve duas tarefas:

**Tarefa 01**
Prever admissão na UTI de casos confirmados de COVID-19.
Com base nos dados disponíveis, é viável prever quais pacientes precisarão de suporte em unidade de terapia intensiva?
O objetivo é fornecer aos hospitais terciários e trimestrais a resposta mais precisa, para que os recursos da UTI possam ser arranjados ou a transferência do paciente possa ser agendada.

**Tarefa 02**
Prever NÃO admissão à UTI de casos COVID-19 confirmados.
Com base nos dados disponíveis, é possível prever quais pacientes precisarão de suporte de unidade de terapia intensiva?
O objetivo é fornecer aos hospitais locais e temporários uma resposta boa o suficiente, para que os médicos de linha de frente possam dar alta com segurança e acompanhar remotamente esses pacientes.


###**Desafio 03 -pesquisar aprendizado supervisionado, não supervisionado, por reforço e mais**

**Aprendizado supervisionado:**  Nesse caso, estamos ensinando à máquina como trabalhar com os nossos dados, logo, estamos supervisionando o aprendizado da máquina. 

No aprendizado supervisionado passamos todos os dados para treinar o modelo. Isto é, passamos os dados e os resultados. Com base nisso, o algoritmo descobre uma forma de trabalhar com os dados.

**Aprendizado não supervisionado:** Diferente do aprendizado supervisionado, neste tipo de aprendizado nós não passamos os resultados para ensinar o algoritmo. 

O próprio algoritmo pega o conjunto de dados e descobre como trabalhar com eles. Os tipos de algoritmos mais utilizados nesse tipo de aprendizado são o clustering e as regras de associação. Um exemplo de algoritmo é o K-Means.

**Aprendizado de reforço:** Nele a máquina interage com o ambiente. Essa interação gera um feedback, positivo ou negativo. Caso seja um feedback positivo, a máquina grava que pode realizar essa ação novamente, em caso de feedback negativo, o modelo sabe que não pode executá-lo novamente.

###**Desafio 04 - Pesquisar sobre classificação, regressão e outros problemas que podem ser solucionados com Machine Learning.**

**Regressão e Classificação Linear**
Estes talvez sejam os algoritmos mais simples em aprendizagem de máquina. Você dispõe de características (ou atributos) x1, x2, …, xn sobre objetos (matriz A) e os seus rótulos (vetor b). O objetivo é encontrar um conjunto de pesos e viés ótimo para estas características de acordo com alguma função de custo, como por exemplo MSE (mean squared error) ou MAE (mean absolute error) para regressão. 

**Regressão Logística**
Este algoritmo calcula a probabilidade do objeto pertencer a cada classe. Por isso, a função de custo pode ser a diferença entre a probabilidade predita e o rótulo 0 ou 1, e calcular a média para todos os objetos como é feito na regressão linear.


**O que eu encontrei e li sobre o assunto:**

[Conhecendo os tipos de aprendizado de máquina](https://www.alura.com.br/artigos/conhecendo-os-tipos-de-aprendizado-de-maquina?gclid=CjwKCAjw_o-HBhAsEiwANqYhp5XVw62LiKsMv2KgPYccTDvoveSMoHMVBlIy2-un2nVIY0xCZ3C3rBoCQ8AQAvD_BwE)

[Algoritmos de Aprendizagem de Máquina: qual deles escolher?](https://medium.com/machina-sapiens/algoritmos-de-aprendizagem-de-m%C3%A1quina-qual-deles-escolher-67040ad68737)

[Ebook  Machine Learning for Dummies, IBM](https://www.ibm.com/downloads/cas/GB8ZMQZ3)

[Simplificando Algoritmos de Machine Learning](https://medium.com/data-hackers/simplificando-as-categorias-de-algoritmos-de-machine-learning-191ac9dc8ddf)

[10 Machine Learning Algorithms You need to Know](https://towardsdatascience.com/10-machine-learning-algorithms-you-need-to-know-77fb0055fe0)







###**Desafio 05 - Qual o motivo de jogar fora o ICU do dado X? O que acontece se estiver no x? Pq seria bom/ruim?**


Testar em cima dos dados de treino cria um viés pra quem está sendo treinado, por isso os dados de treino e teste têm que estar distribuídos de forma adequada. Se os dados de ICU estiverem em x, as previsões terão a acurácia muito alto devido ao enviesamento dos dados. 

###**Desafio 06 - Pesquisar sobre ética em data science/ciência de dados e inteligência artificial.**

A ciência de dados oferece grandes oportunidades para melhorar a vida privada e pública, bem como o nosso meio ambiente.Porém, essas oportunidades também estão associadas a desafios éticos significativos. O uso extensivo de cada vez mais dados - muitas vezes pessoais, senão confidenciais e a crescente dependência de algoritmos para analisá-los a fim de moldar escolhas e tomar decisões, também como a redução gradual do envolvimento humano ou mesmo a supervisão de muitos processos automáticos, colocam questões prementes de justiça, responsabilidade e respeito aos direitos humanos, entre outros.

A [Lei Geral de Proteção de Dados Pessoais (LGPD)](http://www.planalto.gov.br/ccivil_03/_ato2015-2018/2018/lei/L13709.html) entrou em vigor em setembro de 2020 no Brasil e regulamenta qualquer atividade que envolva utilização de dados pessoais. É aplicável a toda operação de tratamento de dados realizada por pessoa física ou jurídica, pública ou privada, independente do meio e de onde os dados estejam localizados. Na prática, a LGPD busca dar às pessoas mais controle sobre quais dados são fornecidos às empresas e como eles são utilizados.

Com a LGPD, regulamenta-se o controle dos dados pelos usuários, tornando obrigatório que as empresas informem quais informações são coletadas e como elas serão usadas. Assim, as pessoas têm a opção de autorizar ou não o compartilhamento.

Algoritmos de inteligência artificial são beneficiados pela maior quantidade de dados possível para a detecção de padrões, e é desejável mantê-los armazenados, já que os padrões históricos podem servir como base para melhores decisões das IAs. É aí que está o grande desafio da LGPD: é preciso limitar a coleta de dados àqueles necessários para os objetivos declarados, reduzindo também o tempo de armazenamento deles, e declarando todas essas questões de forma transparente aos usuários.

Uma vez que os dados foram coletados, a empresa precisa ser capaz de alterar ou apagar os dados do usuário caso ele faça essa solicitação. Portanto, para estar em conformidade com a LGPD, os dados usados em sistemas de inteligência artificial precisarão ser identificáveis e acessíveis a nível individual.

Além disso, é possível que a proteção de dados incentive o desenvolvimento de inteligências artificiais que não dependem necessariamente dos dados compartilhados pelos próprios usuários, como é o caso dos sistemas de reconhecimento de imagem, nos quais a LGPD terá menor influência.

Tecnologias de inteligência artificial não serão prejudicadas, se administradas seguindo o que determina a lei: informar os dados coletados e o tempo de armazenamento, certificar-se de que sejam identificáveis e removíveis, e explicar o que será feito com as informações. 

> Referências:
> - [Proteção de dados e Inteligência Artificial: desafio da LGPD](https://neuralmind.ai/blog/protecao-de-dados-e-inteligencia-artificial-o-desafio-da-lgpd/)
> - [What is data ethics?](https://royalsocietypublishing.org/doi/10.1098/rsta.2016.0360#)




###**Desafio 07 - Decida quanto você vai escolher para treino e quanto para teste? Quanto as pessoas utilizam e qual o motivo?**

Normalmente as pessoas utilizam entre 70% - 80% dos dados no treino e o restante para teste.

Treinar o modelo acima disso pode gerar o problema de *Overfitting*(ocorre quando o modelo se adaptou muito bem aos dados que foi treinado, mas não generaliza bem para novos dados, seria como se ele tivesse decorado mas não aprendido de fato).

> Referências: 
- https://cursos.alura.com.br/forum/topico-duvida-sobre-treino-e-teste-148234
- https://towardsdatascience.com/finally-why-we-use-an-80-20-split-for-training-and-test-data-plus-an-alternative-method-oh-yes-edc77e96295d




###**Desafio 08 - Olhe a documentação da LogisticRegression... o que tem de parâmetro em comum com o train_test_split? O que isso significa para nós?**

Ambos possuem o parâmetro 'random_state', que é reponsável pode definir a semente que irá gerar os números aletórios afim de tomar decisões aleatoriamente:

- `class sklearn.linear_model.LogisticRegression(penalty='l2', *, dual=False, tol=0.0001, C=1.0, fit_intercept=True, intercept_scaling=1, class_weight=None, random_state=None, solver='lbfgs', max_iter=100, multi_class='auto', verbose=0, warm_start=False, n_jobs=None, l1_ratio=None)`


- `sklearn.model_selection.train_test_split(*arrays, test_size=None, train_size=None, random_state=None, shuffle=True, stratify=None) `

Portanto, devemos ter precaução ao definir semente aletória, pois sempre que rodarmos essas funções o correto é definir uma semente novamente.


> **Documentações:**
>
>[LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)
>
>[train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)




###**Super desafio: não usar backfill da janela após a pessoa ter ido pra UTI.**

Escolhi primeiro usar a função `prepare_window `e depois fazer o backfill, pois fazendo ao contrário pode ser que alguns preenchimentos tenham ocorrido depois da pessoa entrar na UTI.


In [455]:
dados = pd.read_excel('https://github.com/vqrca/bootcamp_alura_aulas/blob/main/Modulo_4/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true')
dados

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                            0            1  ...       0-2    0
1                            0            1  ...       2-4    0
2                            0            1  ...       4-6    0
3                            0            1  ...      6-12    0
4                            0            1  ...  ABOVE_12    1
...                        ...          ...  ...       ...  ...
1920                       384            0  ...       0-2    0
1921                       384            0  ...       2-4    0
1922                       384            0  ...       4-6    0
1923                       384            0  ...      6-12    0
1924                       384            0  ...  ABOVE_12    0

[1925 rows x 231 columns]

Função para pegar somente a primeira janela:

In [456]:
def prepare_window(rows):
    if(np.any(rows["ICU"])):
        rows.loc[rows["WINDOW"]=="0-2", "ICU"] = 1
    return rows.loc[rows["WINDOW"] == "0-2"]

"""Essa função recebe as linhas (rows), que são as linhas de um paciente e se qualquer
linha ICU do paciente for = 1  vamos pegar a primeira linha desse paciente e marcar que
ele foi para UTI. Então, a linha onde está a janela 0-2 será localizada e retornada através de 
rows.loc[rows["WINDOW"] == "0-2" """

dados = dados.groupby("PATIENT_VISIT_IDENTIFIER").apply(prepare_window)
dados.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
PATIENT_VISIT_IDENTIFIER                                            ...            
0                        0                          0            1  ...    0-2    1
1                        5                          1            1  ...    0-2    1
2                        10                         2            0  ...    0-2    1
3                        15                         3            0  ...    0-2    0
4                        20                         4            0  ...    0-2    0

[5 rows x 231 columns]

Identificando as colunas contínuas, categóricas e os dados de saída.

Aplicação do método `fillna`:

In [457]:
def preenche_tabela(dados):
    features_continuas_colunas = dados.iloc[:, 13:-2].columns
    features_continuas = dados[features_continuas_colunas].fillna(method='bfill').fillna(method='ffill')
    features_categoricas = dados.iloc[:, :13]
    saida = dados.iloc[:, -2:] #:, todas as linhas
    dados_finais = pd.concat([features_categoricas, features_continuas, saida], ignore_index=True,axis=1) #definir eixo=1 
    dados_finais.columns = dados.columns
    return dados_finais

In [458]:
dados_limpos = preenche_tabela(dados)
dados_limpos

PATIENT_VISIT_IDENTIFIER  ...  ICU
PATIENT_VISIT_IDENTIFIER                                 ...     
0                        0                            0  ...    1
1                        5                            1  ...    1
2                        10                           2  ...    1
3                        15                           3  ...    0
4                        20                           4  ...    0
...                                                 ...  ...  ...
380                      1900                       380  ...    1
381                      1905                       381  ...    0
382                      1910                       382  ...    1
383                      1915                       383  ...    0
384                      1920                       384  ...    0

[385 rows x 231 columns]

Eliminando os dados nulos que restaram:

In [459]:
dados_limpos = dados_limpos.dropna()
dados_limpos.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
PATIENT_VISIT_IDENTIFIER                                            ...            
0                        0                          0            1  ...    0-2    1
1                        5                          1            1  ...    0-2    1
2                        10                         2            0  ...    0-2    1
3                        15                         3            0  ...    0-2    0
4                        20                         4            0  ...    0-2    0

[5 rows x 231 columns]

In [460]:
linhas, colunas = dados_limpos_features.shape
print(f'Os dados limpos após a remoção das colunas indesejadas, contém {linhas} pacientes e {colunas} colunas.')

Os dados limpos após a remoção das colunas indesejadas, contém 352 pacientes e 61 colunas.


Testando os modelos:

In [330]:
x_columns = dados_limpos.describe().columns

In [331]:
y = dados_limpos["ICU"]
x = dados_limpos[x_columns].drop(["ICU"], axis=1)

In [417]:
np.random.seed(79230)
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y, test_size=0.3)

DummyClassifier

In [418]:
modelo = DummyClassifier()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.5603448275862069

LogisticRegression

In [419]:
modelo = LogisticRegression(max_iter=10000)
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.6982758620689655

###**Desafio 09 - Arranque fora as colunas que não devem ser utilizadas.**

In [433]:
dados_limpos.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
PATIENT_VISIT_IDENTIFIER                                            ...            
0                        0                          0            1  ...    0-2    1
2                        10                         2            0  ...    0-2    1
3                        15                         3            0  ...    0-2    0
4                        20                         4            0  ...    0-2    0
5                        25                         5            0  ...    0-2    0

[5 rows x 231 columns]

As colunas retiradas serão aquelas que contêm os valores 'MIN', 'MAX', 'MEDIAN' e 'DIFF', pois apenas a média de cada *feature* pode ser suficiente para tornar os dados mais limpos:

In [434]:
nomenclaturas_removidas = ['MIN', 'MAX', 'MEDIAN', 'DIFF']

lista = []
for nomenclatura in nomenclaturas_removidas:
  for column in dados_limpos.columns:
    if column.endswith(nomenclatura):
      lista.append(column)

dados_limpos_features = dados_limpos.drop(lista,axis=1)

Agora temos somente as médias dos exames:

In [435]:
dados_limpos_features.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
PATIENT_VISIT_IDENTIFIER                                            ...            
0                        0                          0            1  ...    0-2    1
2                        10                         2            0  ...    0-2    1
3                        15                         3            0  ...    0-2    0
4                        20                         4            0  ...    0-2    0
5                        25                         5            0  ...    0-2    0

[5 rows x 63 columns]

Tratamento da coluna `AGE_PERCENTIL`, tranformando str em int64:

In [436]:
dados_limpos_features['AGE_PERCENTIL'] = dados_limpos_features['AGE_PERCENTIL'].str.replace('th', '')
dados_limpos_features['AGE_PERCENTIL'] = dados_limpos_features['AGE_PERCENTIL'].str.replace('Above ', '')
dados_limpos_features['AGE_PERCENTIL'] = dados_limpos_features['AGE_PERCENTIL'].astype('int64')

Remoção das colunas 'PATIENT_VISIT_IDENTIFIER' e 'GENDER': 

In [437]:
dados_limpos_features =dados_limpos_features.drop(columns=['PATIENT_VISIT_IDENTIFIER','GENDER'])
dados_limpos_features = dados_limpos_features.reset_index(drop=True)
dados_limpos_features.head()

AGE_ABOVE65  AGE_PERCENTIL  ...  WINDOW  ICU
0            1             60  ...     0-2    1
1            0             10  ...     0-2    1
2            0             40  ...     0-2    0
3            0             10  ...     0-2    0
4            0             10  ...     0-2    0

[5 rows x 61 columns]

In [438]:
linhas, colunas = dados_limpos_features.shape
print(f'Os dados limpos após a remoção das colunas indesejadas, contém {linhas} pacientes e {colunas} colunas.')

Os dados limpos após a remoção das colunas indesejadas, contém 352 pacientes e 61 colunas.


In [439]:
dados_limpos_features.describe()

AGE_ABOVE65  AGE_PERCENTIL  ...  OXYGEN_SATURATION_DIFF_REL         ICU
count   352.000000     352.000000  ...                  352.000000  352.000000
mean      0.457386      51.818182  ...                   -0.985390    0.463068
std       0.498890      27.395820  ...                    0.033025    0.499344
min       0.000000      10.000000  ...                   -1.000000    0.000000
25%       0.000000      30.000000  ...                   -1.000000    0.000000
50%       0.000000      50.000000  ...                   -1.000000    0.000000
75%       1.000000      80.000000  ...                   -0.980333    1.000000
max       1.000000      90.000000  ...                   -0.732438    1.000000

[8 rows x 60 columns]

Agora vamos testar os modelos após a retirada dessas colunas:

In [440]:
x_columns = dados_limpos_features.describe().columns

In [441]:
y = dados_limpos_features["ICU"]
x = dados_limpos_features[x_columns].drop(["ICU"], axis=1)

In [442]:
np.random.seed(79230)

In [443]:
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y, test_size=0.3)

DummyClassifier

In [444]:
modelo = DummyClassifier()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.5566037735849056

LogisticRegression

In [445]:
modelo = LogisticRegression(max_iter=10000)
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.7358490566037735

**Segundo teste - sem remover 'MIN', 'MAX', 'MEDIAN' e 'DIFF':**

In [446]:
dados_limpos_features_2 = dados_limpos
dados_limpos_features_2.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ... WINDOW  ICU
PATIENT_VISIT_IDENTIFIER                                            ...            
0                        0                          0            1  ...    0-2    1
2                        10                         2            0  ...    0-2    1
3                        15                         3            0  ...    0-2    0
4                        20                         4            0  ...    0-2    0
5                        25                         5            0  ...    0-2    0

[5 rows x 231 columns]

In [447]:
dados_limpos_features_2['AGE_PERCENTIL'] = dados_limpos_features_2['AGE_PERCENTIL'].str.replace('th', '')
dados_limpos_features_2['AGE_PERCENTIL'] = dados_limpos_features_2['AGE_PERCENTIL'].str.replace('Above ', '')
dados_limpos_features_2['AGE_PERCENTIL'] = dados_limpos_features_2['AGE_PERCENTIL'].astype('int64')

In [448]:
dados_limpos_features_2 =dados_limpos_features_2.drop(columns=['PATIENT_VISIT_IDENTIFIER','GENDER'])
dados_limpos_features_2 = dados_limpos_features_2.reset_index(drop=True)
dados_limpos_features_2.head()

AGE_ABOVE65  AGE_PERCENTIL  ...  WINDOW  ICU
0            1             60  ...     0-2    1
1            0             10  ...     0-2    1
2            0             40  ...     0-2    0
3            0             10  ...     0-2    0
4            0             10  ...     0-2    0

[5 rows x 229 columns]

In [449]:
dados_limpos_features_2.describe()

AGE_ABOVE65  AGE_PERCENTIL  ...  OXYGEN_SATURATION_DIFF_REL         ICU
count   352.000000     352.000000  ...                  352.000000  352.000000
mean      0.457386      51.818182  ...                   -0.985390    0.463068
std       0.498890      27.395820  ...                    0.033025    0.499344
min       0.000000      10.000000  ...                   -1.000000    0.000000
25%       0.000000      30.000000  ...                   -1.000000    0.000000
50%       0.000000      50.000000  ...                   -1.000000    0.000000
75%       1.000000      80.000000  ...                   -0.980333    1.000000
max       1.000000      90.000000  ...                   -0.732438    1.000000

[8 rows x 228 columns]

In [450]:
x_columns = dados_limpos_features_2.describe().columns

In [451]:
y = dados_limpos_features_2["ICU"]
x = dados_limpos_features_2[x_columns].drop(["ICU"], axis=1)

In [452]:
np.random.seed(79230)
x_train, x_test, y_train, y_test = train_test_split(x, y, stratify=y, test_size=0.3)

DummyClassifier

In [453]:
modelo = DummyClassifier()
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.5566037735849056

LogisticRegression

In [454]:
modelo = LogisticRegression(max_iter=10000)
modelo.fit(x_train, y_train)

y_prediction = modelo.predict(x_test)
accuracy_score(y_test, y_prediction)

0.6981132075471698

###**Desafio 10 - Essa aula não foi simples, repasse por ela e garanta que entendeu passo a passo.**

✅ Aula foi repassada